crates/re_viewer/build.rs

```diff
@@ -1,4 +1,233 @@
+//! This version is resolved according to the current environment:
+//!
+//! If the `CI` env var is set + the branch name is not `main`, then:
+//! - On any `release-x.y.z` branch, the version is `version/x.y.z`
+//! - On any other branch, the version is `commit/$COMMIT_SHORT_HASH`
+//!
+//! Otherwise, the version is `version/nightly`. This means local builds,
+//! and builds on `main` point to `version/nightly`.
+
+use std::path::Path;
+
+use xshell::cmd;
+use xshell::Shell;
+
+type AnyError = Box<dyn std::error::Error + Send + Sync + 'static>;
+type Result<T, E = AnyError> = std::result::Result<T, E>;
+
+#[derive(Debug)]
+struct Error(String);
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&self.0)
+    }
+}
+
+impl std::error::Error for Error {}
+
+macro_rules! error {
+    ($lit:literal) => (Error($lit.to_owned()));
+    ($($tt:tt)*) => (Error(format!($($tt)*)));
+}
+
+macro_rules! bail {
+    ($lit:literal) => (return Err(error!($lit)));
+    ($($tt:tt)*) => (return Err(error!($($tt)*).into()));
+}
+
+fn git_branch_name(sh: &Shell) -> Result<String> {
+    Ok(String::from_utf8(
+        cmd!(sh, "git rev-parse --abbrev-ref HEAD").output()?.stdout,
+    )?)
+}
+
+fn git_short_hash(sh: &Shell) -> Result<String> {
+    let full_hash = String::from_utf8(cmd!(sh, "git rev-parse HEAD").output()?.stdout)?;
+    Ok(full_hash.trim()[0..7].to_string())
+}
+
+fn parse_release_version(branch: &str) -> Option<String> {
+    // release-\d+.\d+.\d+
+
+    let branch = branch.strip_prefix("release-")?;
+
+    let mut version_parts = branch.split('.');
+    let major = version_parts.next()?.parse::<u8>().ok()?;
+    let minor = version_parts.next()?.parse::<u8>().ok()?;
+    let patch = version_parts.next()?.parse::<u8>().ok()?;
+
+    if version_parts.next().is_some() {
+        return None;
+    }
+
+    Some(format!("{major}.{minor}.{patch}"))
+}
+
+#[derive(serde::Deserialize)]
+struct Frontmatter {
+    title: Option<String>,
+    tags: Option<Vec<String>>,
+    description: Option<String>,
+    thumbnail: Option<String>,
+    thumbnail_dimensions: Option<[u64; 2]>,
+    #[serde(default)]
+    demo: bool,
+}
+
+#[derive(serde::Serialize)]
+struct ManifestEntry {
+    name: String,
+    title: String,
+    description: String,
+    tags: Vec<String>,
+    demo_url: String,
+    rrd_url: String,
+    thumbnail: Thumbnail,
+}
+
+impl ManifestEntry {
+    fn new(example: Example, base_url: &str) -> Result<Self> {
+        macro_rules! get {
+            ($e:ident, $f:ident) => {
+                match $e.readme.$f {
+                    Some(value) => value,
+                    None => bail!("{:?} is missing field {:?}", $e.name, stringify!($f)),
+                }
+            };
+        }
+
+        let thumbnail_dimensions = get!(example, thumbnail_dimensions);
+
+        Ok(Self {
+            title: get!(example, title),
```

Comment:
Why use a macro to unwrap the `title: Option<String>` instead of using `title: String` directly?

Reply:
I think I just wanted the parsing from YAML to not fail early, so that I could check for the presence of `build_args`. I just defaulted all the fields instead.